system.dataset.leftJoin

JosephTLockwood opened the pull request:
Allows the user to left join two datasets. It could also be implemented as a vlookup.
```
leftJoin("dataset", "dataset2", "columnIndex", "columnIndex2")

leftJoin.desc=Performs a left join on two datasets, returning a new dataset.
leftJoin.param.dataset=The first dataset. Must not be null.
leftJoin.param.dataset2=The second dataset. Must not be null.
leftJoin.param.columnIndex=Column index in the first dataset to join on.
leftJoin.param.columnIndex2=Column index in the second dataset to join on.
leftJoin.returns=A new dataset joined on the specified columns of dataset1 and dataset2.
```
#51 
#27

Function descriptions and a test case are included. Feel free to change how you want.

JosephTLockwood commented:
Once again, I don't know how to use Github fully, so not sure why it is showing all commits.

JosephTLockwood commented:
The only thing left should be variable name cleanup. I figure you had some naming convention, so I'll let you handle that.

paul-griffith commented:
Interesting. So this is acting as a horizontal concatenation when the provided column indices line up...

JosephTLockwood commented:
To me, ideally, this function would work the same as SQL's left, right, outer, and inner join. You would be able to add the logic of join on `>` ,`<`, `=`, or `like`. This seemed like a little more work than I had time for, so I figured I could do something similar to DataFrame.merge from the pandas' library. They have many nice options that I have not added yet. 

I believe the best approach if you wanted to include or exclude other columns, would be a different function. Although it is not the fastest as you are processing the data twice I think it would make it the cleanest. Something like this maybe? (Code has not been tested)

```
    @Suppress("unused")
    @ScriptFunction(docBundlePrefix = "DatasetExtensions")
    @KeywordArgs(
        names = ["dataset", "columnsToSplit"],
        types = [Dataset::class, Array<Array<Int>>::class],
    )
    fun splitter(args: Array<PyObject>, keywords: Array<String>): Array<Dataset?> {
        val parsedArgs = PyArgParser.parseArgs(
            args,
            keywords,
            arrayOf("dataset", "columnsToSplit"),
            arrayOf(Dataset::class.java, PyObject::class.java),
            "splitter",
        )
        val dataset = parsedArgs.requirePyObject("dataset").toJava<Dataset>()
        val columnsToSplit = parsedArgs.requirePyObject("columnsToSplit").toJava<Array<Array<Int>>>()
        val datasetSplit = Array<Dataset?>(columnsToSplit.size) { null }

        var currentDataset = 0
        for (newDataSets in columnsToSplit) {
            val columnNames = StringBuilder(newDataSets.size)
            val columnTypes = mutableListOf<Class<*>>()

            newDataSets.forEachIndexed { index, column ->
                columnNames.append(dataset.columnNames[column])
                columnTypes.add(dataset.columnTypes[column])
                if (index != newDataSets.lastIndex) {
                    columnNames.append(", ")
                }
            }

            val builder = DatasetBuilder.newBuilder()
                .colNames(columnNames.toString())
                .colTypes(columnTypes)

            for (row in dataset.rowIndices) {
                val listToAppend = newDataSets.map { dataset[row, it] }.toTypedArray()
                builder.addRow(*listToAppend)
            }

            datasetSplit[currentDataset] = builder.build()
            currentDataset++
        }

        return datasetSplit
    }
```

JosephTLockwood commented:
Added splitter. I also added left, right, inner, and outer. Most likely, this should be condensed to one function where the user chooses a type like in pandas' library. I am also thinking of adding vlookup and hlookup options to this. It would auto-use the trim function, which would be slower than having it built into the functions but would allow for more flexibility.

paul-griffith commented:
I have a clean git history available here:
https://github.com/paul-griffith/ignition-extensions/tree/leftjoin-test

But I can't force-push to this PR/your repository to clean up git on this PR. If you check out that branch, you should be able to create a new PR with clean git history.
